010-provider-architecture/002-final/lib/main.dart

```diff
@@ -1,28 +1,89 @@
+import 'dart:async';
+
 import 'package:flutter/material.dart';
 import 'package:provider/provider.dart';
 import 'package:provider_architecutre/core/services/authentication_service.dart';
-import 'package:provider_architecutre/locator.dart';
+import 'package:provider_architecutre/core/viewmodels/login_model.dart';
 import 'package:provider_architecutre/ui/router.dart';
+import 'package:provider_architecutre/ui/views/home_view.dart';
+import 'package:provider_architecutre/ui/views/login_view.dart';
 
 import 'core/models/user.dart';
+import 'core/services/api.dart';
+import 'core/viewmodels/home_model.dart';
 
 void main() {
-  setupLocator();
   runApp(MyApp());
 }
 
 class MyApp extends StatelessWidget {
   @override
   Widget build(BuildContext context) {
-    return  StreamProvider<User>(
-          initialData: User.initial(),
-          builder: (context) => locator<AuthenticationService>().userController,
-          child: MaterialApp(
-            title: 'Flutter Demo',
-            theme: ThemeData(),
-            initialRoute: '/login',
-            onGenerateRoute: Router.generateRoute,
-          ),
+    return MultiProvider(
```

Comment:
First of all thank you so much for taking the time to write out this example. I really appreciate it. It gives me insights and some other ideas on how I might use it. But this implementation is definitely not for me.

This is exactly what I was talking about with the verbosity of using Provider only. It can definitely be done, I did it this way and as I said over on reddit. Things got way too messy. This is what I consider to be messy (in terms of too much code for no additional gain over a much cleaner approach). 


The class went from 28 lines to 89. Remember that I want to teach the ideas of setting up architectures, if someone wants to implement this afterwards I think they can. Showing a beginner all of this in my tutorials would probably scare some of them away and make them look for a better approach, which is what happened to me. I personally will not be using this method of injection because of how long and verbose it is. I like code that's easier to read and manage, it makes my job more fun and easier to share with other developers. I know in the Functional community and the developers moving away from OOP this is a laughable concept, and in that mindset I do agree. But at the end of the day the dev is the one spending hours in a code base, something as verbose as this (even though it has less dependencies) is not something I want to look at for 300+ hours and explain to new developers coming in or when handing over code bases.

Reply:
Bear in mind that I deleted both `locator.dart` and `base_view.dart`. I also handled the dispose of everything, which you forgot. 
And the once the todos are applies, each proxyprovider will loose 2-3 lines of code. 

So overall the code size if the same. 

Also, you can extract the MultiProvider into the original locator file if you want to.

Reply:
Same goes for all the models that I wrapped into a custom getter to make the field immutable.
Optional. It just prevents silly mistakes where others think they can modify it.

Reply:
I get all of that, and I can see that it **can** be done with Provider only. I just don't find it as neat as 

```dart
locator.register(MyObject()); 
```

That's one line compared to 

```dart
ProxyProvider<DependentService, MyObject>(
          builder: (_, dependentService, previous) =>
              (previous ?? MyObject())..dependentService= dependentService,
          dispose: (_, obj) => obj.dispose(),
        ),
```

Although Kudos for the disposing, I did not show that in this tutorial and it can be removed ofcourse to reduce. But again, reading the top line is easier than reading the bottom line. I also still don't see a clear advantage over using provider only over get_it. They're both doing the same thing, one makes your code easier to understand (in my opinion) and the other doesn't.

Reply:
- auto updates of `MyObject` dependencies
- auto updates of anything that depends on `MyObject`
- overridability (a part of the screen can use a different `MyObject` if needed)
- control over which parts can access `MyObject` and which cannot. 
- clear hooks on where we need to instantiate the object, and where we should dispose of it.


But I'll think of a way to improve the readability of such a mutable `ProxyProvider`.

Maybe something like that:

```dart
ProxyProvider<Dependency, Output>(
  initialBuilder: (_) => Output(),
  builder: (_, dependency, output) => output..dependency = dependency;
  dispose: (_, output) => output.dispose(),
)
```

Is that better?

Reply:
It is a lot better, really like the idea. I'm looking forward the merge of the ProxyProvider to start using provider. Thanks you both for the discussion, as it really let me learn a lot of things. This example was really good and complete. 

In relation with Provider and the changes you did in the other pull request, some providers names feel weird, such as ChangeNotifierProxyProvider2. Another thing that scares at first sight is the large number of different providers available. I hope the differences between them will be well documented.

Reply:
I was hoping we can all learn from the discussion. I'm happy to hear you gained some insight from it.

Reply:
@yaymalaga

`ChangeNotifierProxyProvider` vs `ProxyProvider` has the same relationship than `ChangeNotifierProvider` vs `Provider`.

The only difference between `ChangeNotifierProxyProvider` and `ChangeNotifierProvider` is that the latter has no dependency on other providers, while the former does.

And then the `2` at the end of `ChangeNotifierProxyProvider2` is the number of dependencies this proxy-provider has.
That `2` is sadly necessary because there are no variadic/optional type-parameters in Dart.


Is that clear enough?
That's a very important point you brought up! 😄

Reply:
@FilledStacks yeah, thanks again for the conversation!
It was insightful in many ways!

Reply:
> And then the `2` at the end of `ChangeNotifierProxyProvider2` is the number of dependencies this proxy-provider has.
> That `2` is sadly necessary because there are no variadic/optional type-parameters in Dart.

Totally clear! If the 2 is just the number of dependencies, then it is not a big deal. Thanks!